File: src/tools/why.ts
import { z } from "zod";
import type { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import type { GitRepo } from "../git";
import { parseFact } from "../facts";
import type { Frontmatter } from "../facts";
import { log } from "../logger";

const WhyInput = z.object({
  file: z.string(),
});

type WhyInputType = z.infer<typeof WhyInput>;

interface WhyResult {
  fact: {
    file: string;
    frontmatter: Frontmatter;
    title: string;
    body: string;
  };
  learning_moment: {
    tag: string;
    date: string;
    siblings: Array<{ file: string; title: string; commit: string }>;
  };
  refs: string[];
  history: Array<{ commit: string; date: string; message: string }>;
}

export async function whyHandler(
  repo: GitRepo,
  input: WhyInputType
): Promise<WhyResult> {
  log.info(`why: file="${input.file}"`);
  const syncResult = await repo.sync();
  if (syncResult.conflict) {
    throw new Error(`Merge conflict from origin/main. Conflicting files: ${syncResult.conflict.files.join(", ")}. Resolve with knomit_update then retry.`);
  }

  const history = await repo.log(input.file);

  if (history.length === 0) {
    return {
      fact: {
        file: input.file,
        frontmatter: {
          domain: [],
          confidence: 0,
          sources: 0,
          entities: [],
          refs: [],
        },
        title: "",
        body: "",
      },
      learning_moment: { tag: "", date: "", siblings: [] },
      refs: [],
      history: [],
    };
  }

  // Read and parse the fact
  const raw = await repo.readFile(input.file);
  const parsed = parseFact(raw);

  // Find the learning moment: get the earliest commit, find tags containing it
  const earliestCommit = history[history.length - 1]!;
  const tags = await repo.tagsContaining(earliestCommit.commit);
  const learnTag = tags.find((t) => t.startsWith("learn/")) ?? "";

  // Find siblings from the same learning moment
  let siblings: Array<{ file: string; title: string; commit: string }> = [];
  let tagDate = earliestCommit.date;

  if (learnTag) {
    const commits = await repo.commitsBetweenTags(learnTag);
    siblings = commits
      .filter((c) => c.file !== input.file)
      .map((c) => ({
        file: c.file,
        title: c.message.replace(/^learn: /, ""),
        commit: c.commit,
      }));
  }

  log.info(`why: tag="${learnTag}", ${history.length} history entries, ${siblings.length} sibling(s)`);
  return {
    fact: {
      file: input.file,
      frontmatter: parsed.frontmatter,
      title: parsed.title,
      body: parsed.body,
    },
    learning_moment: {
      tag: learnTag,
      date: tagDate,
      siblings,
    },
    refs: parsed.frontmatter.refs,
    history,
  };
}

export function registerWhyTool(server: McpServer, repo: GitRepo): void {
  server.tool(
    "knomit_why",
    "Explain the provenance of a fact: when it was learned, what learning moment it belongs to, and what sibling facts were learned at the same time.",
    WhyInput.shape,
    async (input) => {
      const parsed = WhyInput.parse(input);
      const result = await whyHandler(repo, parsed);
      return {
        content: [{ type: "text" as const, text: JSON.stringify(result) }],
      };
    }
  );
}

File: internal/mcp/why.go
package mcp

import (
	"context"
	"encoding/json"
	"fmt"
	"strings"

	mcpgo "github.com/mark3labs/mcp-go/mcp"
)

// whyTool returns the Tool definition for knomit_why.
func whyTool() mcpgo.Tool {
	return mcpgo.NewTool("knomit_why",
		mcpgo.WithDescription("Show the history and learning moment for a fact file."),
		mcpgo.WithString("file",
			mcpgo.Required(),
			mcpgo.Description("Path to the fact file (e.g. know/topic/fact.md)."),
		),
	)
}

// WhyHandler returns the handler function for knomit_why.
func WhyHandler(gs GitStore, ontologyRoot string) func(context.Context, mcpgo.CallToolRequest) (*mcpgo.CallToolResult, error) {
	return func(ctx context.Context, req mcpgo.CallToolRequest) (*mcpgo.CallToolResult, error) {
		// 1. Sync.
		if _, err := gs.Sync(nil); err != nil {
			return mcpgo.NewToolResultError(fmt.Sprintf("sync error: %v", err)), nil
		}

		// 2. Get file argument.
		file := req.GetString("file", "")
		if file == "" {
			return mcpgo.NewToolResultError("file is required"), nil
		}
		file = normalizePath(ontologyRoot, file)

		// 3. Read and parse the fact file.
		content, err := gs.ReadFile(file)
		if err != nil {
			return mcpgo.NewToolResultError(fmt.Sprintf("read file error: %v", err)), nil
		}
		fact, err := ParseFact(file, content)
		if err != nil {
			return mcpgo.NewToolResultError(fmt.Sprintf("parse fact error: %v", err)), nil
		}

		// 4. Get commit log for this file.
		logEntries, err := gs.Log(file)
		if err != nil {
			return mcpgo.NewToolResultError(fmt.Sprintf("log error: %v", err)), nil
		}

		// 5. Find oldest commit hash.
		var learningMoment interface{}
		if len(logEntries) > 0 {
			oldestEntry := logEntries[len(logEntries)-1]
			tags, tagErr := gs.TagsContaining(oldestEntry.Commit)
			if tagErr == nil {
				// Find first tag starting with "learn/".
				var learnTag string
				var learnDate string
				for _, tag := range tags {
					if strings.HasPrefix(tag, "learn/") {
						learnTag = tag
						// Use the oldest commit's date.
						learnDate = oldestEntry.Date
						break
					}
				}
				if learnTag != "" {
					learningMoment = map[string]interface{}{
						"tag":      learnTag,
						"date":     learnDate,
						"siblings": []interface{}{},
					}
				}
			}
		}

		// 6. Build history output.
		type historyEntry struct {
			Commit  string `json:"commit"`
			Date    string `json:"date"`
			Message string `json:"message"`
		}
		history := make([]historyEntry, len(logEntries))
		for i, e := range logEntries {
			history[i] = historyEntry{
				Commit:  e.Commit,
				Date:    e.Date,
				Message: e.Message,
			}
		}

		// 7. Build fact output.
		factOut := map[string]interface{}{
			"file":  fact.Path,
			"title": fact.Title,
			"body":  fact.Body,
			"frontmatter": map[string]interface{}{
				"domain":     orEmpty(fact.Domain),
				"confidence": fact.Confidence,
				"sources":    fact.Sources,
				"entities":   orEmpty(fact.Entities),
				"refs":       orEmpty(fact.Refs),
			},
		}

		result := map[string]interface{}{
			"fact":             factOut,
			"learning_moment":  learningMoment,
			"refs":             orEmpty(fact.Refs),
			"history":          history,
		}

		out, err := json.Marshal(result)
		if err != nil {
			return mcpgo.NewToolResultError(fmt.Sprintf("marshal error: %v", err)), nil
		}
		return mcpgo.NewToolResultText(string(out)), nil
	}
}
